### PR TITLE
Static Code Analyzer

### PR DESCRIPTION
**Pull request by MatiBrego**
Note: Merge #9 first

### CODE REVIEW

**Review by lucaslovaglio — APPROVED** [review]

File: sca/src/main/kotlin/Report.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    fun print(){
+        if (ruleResults.isEmpty()){
+            println("Static code analysis successful!")
+            return
+        }
+        for (ruleResult in ruleResults) {
+            when(ruleResult){
+                is RuleSuccess -> {}
+                is RuleFailure -> printRuleFailure(ruleResult)
+                is RuleFailures -> printFailures(ruleResult)
```
Comment (JuanSeZ) [inline sca/src/main/kotlin/Report.kt]: Why does it need to be handled separated ? Can't we consider `RuleFailure` as a special case of `RuleFailures`

Comment (MatiBrego) [reply sca/src/main/kotlin/Report.kt]: You mean use the same "printFailures" function?

File: sca/src/main/kotlin/factory/StaticCodeAnalyzerFactory.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.example.provider.StaticCodeAnalyzerConfigurer
+import org.example.provider.PrintLnArgumentNonExpressionRuleConfigurer
+import org.example.provider.VariableNamingRuleConfigurer
+
+class StaticCodeAnalyzerProviderFactory {
+
+    private val ruleConfigurers = mapOf(
+        "variableNamingRule" to VariableNamingRuleConfigurer(),
+        "printLnArgumentNonExpressionRule" to PrintLnArgumentNonExpressionRuleConfigurer()
```
Comment (JuanSeZ) [inline sca/src/main/kotlin/factory/StaticCodeAnalyzerFactory.kt:15]: Just to get it clear , we are not considering the case in which these rules are not required right ?

File: sca/src/test/resources/sca.config.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "variableNamingRule" : "^[a-z]+(?:[A-Z][a-z]*)*$",
+  "printLnArgumentNonExpressionRule" : true
+}
```
Comment (JuanSeZ) [inline sca/src/test/resources/sca.config.json:4]: Why should we enter the regex and not explicitly say camel case or snake case

**Review by JuanSeZ — APPROVED** [review]
LGTM

File: sca/src/main/kotlin/factory/StaticCodeAnalyzerFactory.kt (hunk 2 above)
Comment (MatiBrego) [reply sca/src/main/kotlin/factory/StaticCodeAnalyzerFactory.kt:15]: Yes, this is a factory class for the SCA configurer. What I am doing here is creating a configurer that has a map of the rule name that should go in the config file mapped to the especific configurer for that rule. See the StaticCodeAnalyzerConfigurer class to see how it works

File: sca/src/main/kotlin/Report.kt (hunk 1 above)
Comment (JuanSeZ) [reply sca/src/main/kotlin/Report.kt]: Not really, like isn't a `RuleFailure` just a `RuleFailures` with just one element

File: sca/src/test/resources/sca.config.json (hunk 3 above)
Comment (MatiBrego) [reply sca/src/test/resources/sca.config.json:4]: To make it more extensible. This way there is no need of creating new configurers for each naming convention.

File: sca/src/main/kotlin/Report.kt (hunk 1 above)
Comment (MatiBrego) [reply sca/src/main/kotlin/Report.kt]: Yes you are right, I could change the whole thing to just use RuleFailures.

**Pull request merged**